text/0148-logs-for-crashes.md

```diff
@@ -0,0 +1,218 @@
+
+Links to some implementations:
+
+* Java: [PersistingScopeObserver](https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/cache/PersistingScopeObserver.java)
+* Java: [NdkScopeObserver](https://github.com/getsentry/sentry-java/blob/d21770841d84eddba0636e2082c029db2a158457/sentry-android-ndk/src/main/java/io/sentry/android/ndk/NdkScopeObserver.java)
+* Cocoa: [SentryCrashScopeObserver](https://github.com/getsentry/sentry-cocoa/blob/d8ceea3a0ce99c0dd499d3b9472c87aaf0713d3c/Sources/Sentry/SentryCrashScopeObserver.m)
+* Cocoa: [SentryWatchdogTerminationScopeObserver](https://github.com/getsentry/sentry-cocoa/blob/main/Sources/Sentry/SentryWatchdogTerminationScopeObserver.m)
+
+## Requirements
+
+1. The solution MUST NOT lose any log messages in the event of a crash.
+
+```swift
+logger.trace("Starting database connection")
+
+// The above log must show up in Sentry
+SentrySDK.crash()
+```
+
+1. The solution SHOULD minimize the loss of log messages for watchdog terminations. This MAY vary based on the platform. If the platforms allows it, SDKs SHOULD try to not loose any logs for watchdog terminations.
+2. The solution MUST NOT block the main thread.
+3. The solution MUST be thread-safe.
+4. The solution MUST work for hybrid SDKs.
+5. The solution SHOULD NOT depend on or interfere with the offline caching of envelopes, meaning once we tackle the cache overflow problem with priority queues or whatever, little or no changes SHOULD be required for this solution.
+6. The solution MAY also work for spans and other future telemetry data.
+
+# Options Considered
+
+## A - FIFO Queue With Async IO
+
+The existing [BatchProcessor](https://develop.sentry.dev/sdk/telemetry/spans/batch-processor/) minimizes the number of HTTP requests that SDKs make to Sentry for logs and currently stores the logs in memory. When a crash happens, all these logs are lost.
+
+With this option, the BatchProcessor stores its logs in a thread-safe FIFO queue, residing in an async-safe memory space, allowing the crash reporter to write them to disk when a crash occurs. Furthermore, the BatchProcessor stores logs asynchronously into a file, allowing it to recover after an abnormal termination, for which the crash handler can't run. The BatchProcessor MUST store the logs immidiately to disk after adding them to the FIFO queue. It MUST NOT use a timeout as it does for flushing the logs to Sentry to minimize data loss in case of a watchdog termination.
+
+When the BatchProcessor receives a log, it performs the following steps
+
+1. Put the log into the FIFO queue on the calling thread.
+2. On a background thread, serialize the next log of the FIFO queue and store it in the `batch-processor-cache-file`.
+3. Remove the log from the FIFO queue.
```

Comment:
I guess I'm a bit confused here - why do we want to remove the log from the queue? Or is this queue a different one from what we already have in the [BatchProcessor](https://github.com/getsentry/sentry-cocoa/blob/07f6759e7a19918f0e1d14e12f586b32758dcbc3/Sources/Swift/Tools/SentryLogBatcher.swift#L87C13-L87C24)?

My understanding was that just in addition to what we already have (the existing queue) we would just spin up a task to store the same log in a file. If it crashes and the in-memory logs are lost, we'd use the file on the next launch to send the leftovers.

Reply:
> When a crash occurs, the SDKs write the logs in the FIFO queue to the `log-crash-recover-file`  and send these logs on the next SDK launch. 

@romtsn, the FIFO queue exists to prevent log loss when a crash occurs immediately after logging. 

```swift
logger.trace("Starting database connection")

// The above log must show up in Sentry
SentrySDK.crash()
```

If we don't use the async safe memory and only store it to disk on a BG thread, we can't guarantee that on Cocoa. Can you on Java?

Reply:
Yep, let's figure this out once we add this to the develop docs. Thanks for bringing it up @romtsn.